src/api/app/controllers/status_project_controller.rb

```diff
@@ -1,41 +1,68 @@
 class StatusProjectController < ApplicationController
+  before_action :initialize_caches, only: [:show]
+
   def show
     dbproj = Project.get_by_name(params[:project])
     @packages = ProjectStatus::Calculator.new(dbproj).calc_status
-    find_relationships_for_packages(@packages)
+    find_relationships_for_packages
   end
 
   private
 
+  def initialize_caches
+    @package_hash = {}
+  end
+
   def role_from_cache(role_id)
-    @rolecache[role_id] ||= Role.find(role_id).title
+    @rolecache ||= Hash.new do |h, rid|
+      h[rid] = Role.find(rid).title
+    end
+    @rolecache[role_id]
   end
 
   def user_from_cache(user_id)
-    @usercache[user_id] ||= User.find(user_id).login
+    @usercache ||= Hash.new do |h, uid|
+      h[uid] = User.find(uid).login
+    end
+    @usercache[user_id]
   end
 
   def group_from_cache(group_id)
-    @groupcache[group_id] ||= Group.find(group_id).title
+    @groupcache ||= Hash.new do |h, gid|
+      h[gid] = Group.find(gid).title
+    end
+    @groupcache[group_id]
   end
 
-  def find_relationships_for_packages(packages)
-    @usercache = {}
-    @groupcache = {}
-    @rolecache = {}
-    package_hash = {}
-    packages.each_value do |p|
-      package_hash[p.package_id] = p
-      package_hash[p.develpack.package_id] = p.develpack if p.develpack
+  def package_hash(packages)
+    return @package_hash unless @package_hash.empty?
+    packages.each_value do |pkg|
+      @package_hash[pkg.package_id] = pkg
+      @package_hash[pkg.develpack.package_id] = pkg.develpack if pkg.develpack
     end
-    relationships = Relationship.where(package_id: package_hash.keys).pluck(:package_id, :user_id, :group_id, :role_id)
-    relationships.each do |package_id, user_id, group_id, role_id|
+    @package_hash
+  end
+
+  def relationships_for_package(keys)
+    Relationship.where(package_id: keys).pluck(:package_id, :user_id, :group_id, :role_id)
+  end
+
+  def add_person(user_id, role_id, package_id)
```

Comment:
`package_hash(@packages)[package_id]` is just `Package.find(package_id)`
`user_from_cache(user_id)` is just `User.find(user_id)`
`role_from_cache(rold_id)` is just `Role.find(role_id)`

Just another useless cache on top of the `ActiveRecord` query cache IMHO.

Reply:
I think it was discussed here with @coolo https://github.com/openSUSE/open-build-service/pull/5765

@coolo knows (TM)

Reply:
I wouldn't remove the caching unless we are 100% that it is useless:

```
vpereira@kimura:~/open-build-service> time osc api '/status/project/openSUSE:Factory' 1>/dev/null

real	2m40.098s
user	0m0.326s
sys	0m0.210s
```
```
vpereira@kimura:~/open-build-service> time osc api '/status/project/openSUSE:Factory' 1>/dev/null

real	0m53.210s
user	0m0.376s
sys	0m0.101s
```

Reply:
the cache we're talking about is memorize - so you can't measure its impact without modifying the code. The cache you just measure is the backend filesystem's cache (yes, it's really that sad)

Reply:
@coolo yes, I know. What I wanted to point out is that as it is, is already bad enough.. so to remove the memoization just if we are 100% sure, that there is no gain with the memoization. I would prefer to don't drop it in my PR.

Reply:
I see

Reply:
![](https://static.boredpanda.com/blog/wp-content/uploads/2018/09/1037532352451411968-png__700.jpg)

Reply:
Great quote:+1: But what does this actually mean? Do we know by now if and why this needs to be cached? If the route is heavily used we might rather do some [page](https://guides.rubyonrails.org/caching_with_rails.html#page-caching) or [action](https://guides.rubyonrails.org/caching_with_rails.html#action-caching) caching. If the bottleneck is in the backend we should improve it there, etc.

Are we confident that we optimize the right thing? If not I would rather spend some time investigating this instead of spending time refactoring code we might don't need:-)

Reply:
I have no idea if this is really needed any longer - I'd be just very careful with these generic statements about rails doing the right thing. But the route isn't the most often used routed nor do these 3 functions hurt anyone - so I wonder how much effort you want to put into this investigation.

Reply:
We generally like to understand WTF is going on and record this for the developers who come after us...

Reply:
Ok, I will test it on monday to see if its used

Reply:
> coolo 3 days ago
> I have no idea if this is really needed any longer - I'd be just very careful with these generic statements about rails doing the right thing. 

I am not saying that we should drop it without checking. I am saying that we should have a look and verify that we need it when there is doubt.

> so I wonder how much effort you want to put into this investigation.

It's the second time now within short time that we bump into this. Instead of putting effort into discussing this a third time I would spend an hour and address this.
Beside I find it an odd pattern to act like these are ancient monuments that can not be touched nor understand. 

![2001](https://user-images.githubusercontent.com/968949/45300819-a90ae300-b50f-11e8-8a15-b45f88fdf579.jpg)

Reply:
>  this a third time I would spend an hour and address this.

It was more like 5 hours of work  and 4 days that I'm waiting to merge it :) but, right  I think now we understand it better. The route takes a long time to be processed. To memoize code which is used in a loop, relevant just for an instance, is a valid to pattern. The cache query doesn't invalidate it, but helps. If that's the best pattern to implement cache in controllers, it's a different question.. anyway so now I see three options:

a) drop the PR, keep the memoization as it is
b) accept the PR, keep the memoization at it is but with a better code
c) propose a better solution for caching in this controller, implement it, measure and see if it worth it.

I'm for b) and you?

Reply:
> we should have a look and verify that we need it 
That's exactly what I asked for - and nothing more.

When I say "be careful" does it sound to you like "don't do it"? If so, I'm sorry for the confusion